src/main/java/org/opensearch/security/privileges/PrivilegesEvaluatorImpl.java

```diff
@@ -0,0 +1,758 @@
+                    final AliasMetadata aliasMetadata = aliases.get(alias);
+
+                    if (aliasMetadata != null && aliasMetadata.filteringRequired()) {
+                        filteredAliases.add(aliasMetadata);
+                        log.debug("{} is a filtered alias {}", alias, aliasMetadata.getFilter());
+
+                    } else {
+                        log.debug("{} is not an alias or does not have a filter", alias);
+
+                    }
+                }
+            }
+
+            if (filteredAliases.size() > 1 && ACTION_MATCHER.test(action)) {
+                log.error(
+                    "More than one ({}) filtered alias found for same index ({}). This is currently not supported. Aliases: {}",
+                    filteredAliases.size(),
+                    indexMetaData.getIndex().getName(),
+                    toString(filteredAliases)
+                );
+                return true;
+            }
+        } // end-for
+
+        return false;
+    }
+
+    private boolean checkDocAllowListHeader(User user, String action, ActionRequest request) {
+        String docAllowListHeader = threadContext.getHeader(ConfigConstants.OPENDISTRO_SECURITY_DOC_ALLOWLIST_HEADER);
+
+        if (docAllowListHeader == null) {
+            return false;
+        }
+
+        if (!(request instanceof GetRequest)) {
+            return false;
+        }
+
+        try {
+            DocumentAllowList documentAllowList = DocumentAllowList.parse(docAllowListHeader);
+            GetRequest getRequest = (GetRequest) request;
+
+            if (documentAllowList.isAllowed(getRequest.index(), getRequest.id())) {
+                log.debug("Request {} is allowed by {}", request, documentAllowList);
+                return true;
+            } else {
+                return false;
+            }
+
+        } catch (Exception e) {
+            log.error("Error while handling document allow list: {}", docAllowListHeader, e);
+            return false;
+        }
+    }
+
+    private List<String> toString(List<AliasMetadata> aliases) {
+        if (aliases == null || aliases.size() == 0) {
+            return Collections.emptyList();
+        }
+
+        final List<String> ret = new ArrayList<>(aliases.size());
+
+        for (final AliasMetadata amd : aliases) {
+            if (amd != null) {
+                ret.add(amd.alias());
+            }
+        }
+
+        return Collections.unmodifiableList(ret);
+    }
+
+    private static Map<String, SubjectBasedActionPrivileges> createActionPrivileges(
+        Map<String, RoleV7> pluginIdToRolePrivileges,
+        FlattenedActionGroups staticActionGroups
+    ) {
+        Map<String, SubjectBasedActionPrivileges> result = new HashMap<>(pluginIdToRolePrivileges.size());
+
+        for (Map.Entry<String, RoleV7> entry : pluginIdToRolePrivileges.entrySet()) {
+            result.put(entry.getKey(), new SubjectBasedActionPrivileges(entry.getValue(), staticActionGroups));
```

Comment:
does this require to be thread-safe? also, can it lead to ``ConcurrentModificaitonException``?

Reply:
This is only called in the constructor, so it is strictly single threaded:

https://github.com/opensearch-project/security/blob/65375fd992c7711efee8abf2c468b90cac719ea0/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluatorImpl.java#L155-L202